add docs on how to debug Pulsar Functions (#4485)
apache/pulsar: site2/docs/functions-debugging.md

---
id: functions-debugging

> title: Debugging Pulsar Functions

Author: @Anonymitaet is that correct to change gerunds to base verbs in headings.  Just looking at what others have done:

1) https://aws.amazon.com/getting-started/
2) https://cloud.google.com/gcp/getting-started/

For AWS lambda functions:
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-debugging.html

Perhaps @sijie @merlimat can also chime in on this

> title: Debugging Pulsar Functions

Reviewer: I'm not an english grammar expert, though "Getting Started" seems to be form used everywhere.

> title: Debugging Pulsar Functions

Reviewer: (And from that on, "Debugging functions" sounds correct in the same way)

> title: Debugging Pulsar Functions

Reviewer 2: @jerrypeng 

Grammatically, both gerund and base verb are correct, but given that the heading title plays a huge role in the search results, using the imperative means the heading matches closely the questions entered by users. **This is a user-centered design**.

### Examples

The following companies are using base verbs in headings.

**Top software companies**

* Get started with Microfost Azure: https://azure.microsoft.com/en-us/get-started/)
* Get started with Oracle Database 19c: https://docs.oracle.com/en/database/oracle/oracle-database/19/index.html
* Get started with IBM Cloud: https://www.ibm.com/cloud/get-started
* Get started with Docker: https://www.docker.com/get-started
* Get started with Adobe Creative Cloud: https://helpx.adobe.com/creative-cloud/get-started.html

**Other software companies** 

* Get started with RabbitMQ: https://www.rabbitmq.com/#getstarted
* Get started with Cloud Foundry: https://www.cloudfoundry.org/get-started/
* Get started with Wowza Streaming Cloud: https://www.wowza.com/docs/getting-started-with-wowza-streaming-cloud
* Get started with Firebase: https://firebase.google.com/docs/database/security/quickstart
* Get started with Drake: https://www.drakesoftware.com/service-learning/get-started-with-drake/

### Trend in the technical writing industry and government

* [The #1 Headline Blunder Every Writer Makes](https://entrepreneurshandbook.co/the-1-headline-blunder-every-writer-makes-weve-all-done-it-e1aa3563dff3) explains other reasons for dropping the gerund starter (headlines that start with a gerund are ambiguous, not reader-centric and weak) and propose to chop off the “-ing” and use the base verb to make it an imperative.

* [Are Gerunds in Topic Titles Problematic in Search Results?](https://idratherbewriting.com/2008/02/11/are-gerunds-in-topic-titles-problematic-in-search-results/)

* [Death of the gerund in technical documentation](https://ffeathers.wordpress.com/2013/10/12/death-of-the-gerund-in-technical-documentation/)

* [The decline of the gerund in technical documentation?](https://www.cherryleaf.com/2015/06/the-decline-of-the-gerund-in-technical-documentation/)

> title: Debugging Pulsar Functions

Reviewer 3: @jerrypeng @merlimat Thank you very much for raising this question.
Actually, "debugging functions" / "debug functions", "Getting started" / "Get started" are correct as a task topic title. When we write task topics, we can use verb-based, gerunds, or "how-to" titles, either is OK, as long as we keep consistent.  
Then you might ask why do we change it as "base verb"? Mainly for two reasons.
1. Minimalism writing style: when IBM writers began to advocate minimalism writing in 201x(I'm sorry I forget the exact year) , most writers in other companies begin to accept this style. We write “minimalist,” task-oriented information that quickly meets users’ needs, create effective tasks, and separate tasks from concept and reference topics. In older documentation, task topics are mixed with concepts and references, it is difficult for users to find what they want faster; in new trend, task topics are separated from concept and reference, so users can easily and quickly find what they want. If you read newer IBM docs, they have adopted the new styles. For older version, some are not updated. For example, 
![image](https://user-images.githubusercontent.com/47805623/59318622-31feb700-8cfa-11e9-9a3e-2e08dbafb296.png)
However, in [DITA Best Practices](https://www.amazon.com/DITA-Best-Practices-Roadmap-Architecting/dp/0132480522) published by IBM press, verb for task topics are recommended.
![image](https://user-images.githubusercontent.com/47805623/59318702-7ab67000-8cfa-11e9-9697-1c0b919f5a8d.png)
In Microsoft documentation, they also adopt the new style, see https://docs.microsoft.com. The following is an example from their website: https://docs.microsoft.com/en-us/azure/devops/get-started/index?view=azure-devops
![image](https://user-images.githubusercontent.com/47805623/59318751-a6d1f100-8cfa-11e9-842a-6a7ef118bd2f.png)
Pingcap doucmentation also adopt the new trend, their writers are mainly former IBMers.
https://pingcap.com/docs/dev/how-to/get-started/local-cluster/install-from-binary/
![image](https://user-images.githubusercontent.com/47805623/59319184-4e9bee80-8cfc-11e9-9b89-36cf0a062f52.png)
Concerning AWS doc recommended by @jerrypeng at https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-quick-start.html, we can see that AWS is typically the older docs, mixing tasks with concepts and references. Yet they are also working towards the task-oriented trend, see their task titles:
![image](https://user-images.githubusercontent.com/47805623/59318859-22cc3900-8cfb-11e9-8c8e-630090042ec0.png)
2. Faster in fuzzy search: Using "base-verb" is easier and faster for users to search information they need.

> title: Debugging Pulsar Functions

Author: @Anonymitaet @Jennifer88huang thank you for the detailed responses.  I will reword the titles.

> sidebar_label: Debugging functions

Author: "Debug Functions" is awkward sounding and confusing as a side part title.  Either we keep this as it is or change it to "How to debug Functions"

> sidebar_label: Debugging functions

Reviewer 2: "Debugging functions" is confusing because users may think "they are functions used with debugging" rather than "How to debug Functions".

Example
Microsoft Debugging Functions: https://docs.microsoft.com/en-us/windows/desktop/debug/debugging-functions.

Why "Debug Functions" is confusing?

   public String apply(String input) {
       return String.format("%s!", input);
   }
}
```

The user can write a simple unit test to test this Pulsar function:

```java
@Test
public void testJavaNativeExclamationFunction() {
   JavaNativeExclamationFunction exclamation = new JavaNativeExclamationFunction();
   String output = exclamation.apply("foo");
   Assert.assertEquals(output, "foo!");
}
```

Consequently, if a user has a Pulsar Function that implements the ```org.apache.pulsar.functions.api.Function``` interface:

```java
import org.apache.pulsar.functions.api.Context;
import org.apache.pulsar.functions.api.Function;

public class ExclamationFunction implements Function<String, String> {
   @Override
   public String process(String input, Context context) {
       return String.format("%s!", input);
   }
}
```

The user can write a unit test for this function as well. Just be remember to mock out the ```Context``` parameter.

For example:

```java
@Test
public void testExclamationFunction() {
   ExclamationFunction exclamation = new ExclamationFunction();
   String output = exclamation.process("foo", mock(Context.class));
   Assert.assertEquals(output, "foo!");
}
```

> ## Debugging with localrun mode

Reviewer 3: ```suggestion
## Debug with the localrun mode
```

Reviewer 3: check other similar cases.

Author: @Jennifer88huang 

For AWS lambda functions:
https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-using-debugging.html

> > Please note that there is currenlty only support for this for Pulsar Functions written in Java.  Users will need a Pulsar version > 2.4.9 to do the following. Even though localrun is available in older version of Pulsar, it does not have the functionally to be executed programmatically and run functions as threads.

Author: yes good catch

> > Please note that there is currenlty only support for this for Pulsar Functions written in Java.  Users will need a Pulsar version > 2.4.9 to do the following. Even though localrun is available in older version of Pulsar, it does not have the functionally to be executed programmatically and run functions as threads.

Author: I will change

To test in a more realistic fashion, a Pulsar Function can be run via localrun mode which will launch a instance of the function on your local machine as a thread.

> In this mode, the Pulsar Function can consume and produce actual data to a Pulsar Cluster and mirrors how the function will actually run in a Pulsar Cluster.

Reviewer 3: 'cluster` is not a proper noun, we can use lowercase.

Author: will change

Users can launch his or her function in the following manner:

```java
FunctionConfig functionConfig = new FunctionConfig();
functionConfig.setName(functionName);
functionConfig.setInputs(Collections.singleton(sourceTopic));
functionConfig.setClassName(ExclamationFunction.class.getName());
functionConfig.setRuntime(FunctionConfig.Runtime.JAVA);
functionConfig.setOutput(sinkTopic);

LocalRunner localRunner = LocalRunner.builder().functionConfig(functionConfig).build();
localRunner.start(true);
```

> This allows users to easily debug functions using an IDE.  User's can set break points and manually set through a function to debug will real data.

Reviewer 2: ```suggestion
This allows users to easily debug functions using an IDE. Users can set breakpoints and manually set through a function to debug with real data.
```

Is "will" a typo?

Author: will fix

> The code snippet below illustrates a more complete example of how to use launch a function via localrun programmatically.

Reviewer 3: ```suggestion
The following code example illustrates  how to launch a function via the localrun mode.
```

Author: will modify

```java
public class ExclamationFunction implements Function<String, String> {

   @Override
   public String process(String s, Context context) throws Exception {
       return s + "!";
   }

public static void main(String[] args) throws Exception {
    FunctionConfig functionConfig = new FunctionConfig();
    functionConfig.setName("exclamation");
    functionConfig.setInputs(Collections.singleton("input"));
    functionConfig.setClassName(ExclamationFunction.class.getName());
    functionConfig.setRuntime(FunctionConfig.Runtime.JAVA);
    functionConfig.setOutput("output");

    LocalRunner localRunner = LocalRunner.builder().functionConfig(functionConfig).build();
    localRunner.start(false);
}
```

To use localrun like above programmatically please addd the following dependency:

```xml
<dependency>
   <groupId>org.apache.pulsar</groupId>
   <artifactId>pulsar-functions-local-runner</artifactId>
   <version>${pulsar.version}</version>
</dependency>

```

> For complete code samples please visit [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Reviewer 3: Currently, I haven't found more information at this site, would you store more code samples in the readme file, or in a separate directory? If so, you can update the site address.

> For complete code samples please visit [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Author: @Jennifer88huang the url points to complete code samples not more documentation

> For complete code samples please visit [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Author: @Anonymitaet I don't think you suggestion is correct.  The link is for complete code samples not more information about complete code samples.  I will add a comma after "samples" though.

> For complete code samples please visit [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Reviewer 2: @jerrypeng 

(1) The link to "complete code samples" is **a kind of information** of "complete code samples", everything about a "subject" is a kind of information.

(2) "For more information, see..." and "For more information on/about..., see..." are "fixed formats" to guide users to other places. 

Currently, almost all technical documents including Pulsar guides, Microsoft guides, IBM guides, and other technical style guides use this format, we need to keep consistently.

> For complete code samples please visit [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Author: "For more information about complete code samples, see [LINK]."

The sentence is confusing and awkward.

"For complete code samples, see [LINK]"

The semantics of above sentence is much clearer.

> In the future, we will add native local run support for Pulsar Functions written in other languages.

Reviewer 2: (1) Avoid using the word “I” or “We” in technical writing. 

Technical documents should be written in the third person. We (technical writers) are reporting, not advertising or advocating.

(2) Keep consistent with previous contents.

functionConfig.setRuntime(FunctionConfig.Runtime.JAVA);
functionConfig.setOutput(sinkTopic);

LocalRunner localRunner = LocalRunner.builder().functionConfig(functionConfig).build();
localRunner.start(true);
```

This allows users to easily debug functions using an IDE. Users can set breakpoints and manually step through a function to debug with real data.

The following code snippet is a more complete example on how to programmatically launch a function in localrun mode.

```java
public class ExclamationFunction implements Function<String, String> {

   @Override
   public String process(String s, Context context) throws Exception {
       return s + "!";
   }

public static void main(String[] args) throws Exception {
    FunctionConfig functionConfig = new FunctionConfig();
    functionConfig.setName("exclamation");
    functionConfig.setInputs(Collections.singleton("input"));
    functionConfig.setClassName(ExclamationFunction.class.getName());
    functionConfig.setRuntime(FunctionConfig.Runtime.JAVA);
    functionConfig.setOutput("output");

    LocalRunner localRunner = LocalRunner.builder().functionConfig(functionConfig).build();
    localRunner.start(false);
}
```

> To use localrun like above programmatically please addd the following dependency:

Reviewer 3: ```suggestion
To use a function in localrun mode as above, add the following dependency:
```

```xml
<dependency>
   <groupId>org.apache.pulsar</groupId>
   <artifactId>pulsar-functions-local-runner</artifactId>
   <version>${pulsar.version}</version>
</dependency>

```

> For complete code samples, see [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging)

Reviewer 3: ```suggestion
For complete code samples, see [here](https://github.com/jerrypeng/pulsar-functions-demos/tree/master/debugging).
```
